File: LLD-questions/load-balancer.py
# Server class
class Server:
    def __init__(self, serverId:) -> None:
        self.isHealthy = True
        self.serverId = serverId
    
    def setHealthy(self, healthy):
        self.isHealthy = healthy
    
    def isHealthy(self):
        return self.isHealthy
    
# Load balancer
class LoadBalancer:
    instance = None
    servers = []
    strategy = None
    @staticmethod
    def getSingleton():
        if instance == None:
            instance = LoadBalancer()
        return instance
    
    def addServer(self, server: Server):
        self.servers.append(server)

    def removeServer(self, server):
        self.servers.remove(server)
    
    def loadBalancingStrategy(self, strategy):
        self.strategy = strategy
    
    def getServer(self, request):
        return self.strategy.getServer(self.servers, request)

class LeastConnectionStrategy:
    def __init__(self) -> None:
        self.minconnection = 10

    
    def getServer(self, servers):
        for server in servers: 
            if server.isHealthy():
                connections = self.getConnections(server)
                if (connections < minConnections):
                    minConnections = connections
                    selectedServer = server
        return selectedServer


    def getConnections(self, server):
        return 0           
            
        


if __name__ == "__main__":
    server1 = Server("sever1")
    server2 = Server("sever2")

    LoadBalancer = LoadBalancer.getSingleton()
    LoadBalancer.addServer(server1)
    LoadBalancer.addServer(server2)

    strategy = RoundRobinStrategy()
    LoadBalancer.loadBalancingStrategy(strategy)

    request1 = input()

File: LLD-questions/load-balancer.java
// Server class
class Server {
    private String serverId;
    private boolean isHealthy;
    // Other attributes and methods

    public Server(String serverId) {
        this.serverId = serverId;
        this.isHealthy = true;
    }

    public boolean isHealthy() {
        return isHealthy;
    }

    public void setHealthy(boolean healthy) {
        isHealthy = healthy;
    }

    // Other server operations
}

// LoadBalancer class (Singleton)
class LoadBalancer {
    private static LoadBalancer instance;
    private List<Server> servers;
    private LoadBalancingStrategy strategy;

    private LoadBalancer() {
        this.servers = new ArrayList<>();
    }

    public static LoadBalancer getInstance() {
        if (instance == null) {
            instance = new LoadBalancer();
        }
        return instance;
    }

    public void addServer(Server server) {
        servers.add(server);
    }

    public void removeServer(Server server) {
        servers.remove(server);
    }

    public Server getServer(Request request) {
        return strategy.getServer(servers, request);
    }

    public void setLoadBalancingStrategy(LoadBalancingStrategy strategy) {
        this.strategy = strategy;
    }

    // Other load balancer operations
}

// LoadBalancingStrategy interface
interface LoadBalancingStrategy {
    Server getServer(List<Server> servers, Request request);
}

// RoundRobinStrategy class
class RoundRobinStrategy implements LoadBalancingStrategy {
    private int currentIndex;

    public RoundRobinStrategy() {
        this.currentIndex = 0;
    }

    @Override
    public Server getServer(List<Server> servers, Request request) {
        int totalServers = servers.size();
        if (totalServers == 0) {
            throw new IllegalStateException("No servers available");
        }
        Server server = servers.get(currentIndex);
        currentIndex = (currentIndex + 1) % totalServers;
        return server;
    }
}

// LeastConnectionsStrategy class
class LeastConnectionsStrategy implements LoadBalancingStrategy {
    @Override
    public Server getServer(List<Server> servers, Request request) {
        int minConnections = Integer.MAX_VALUE;
        Server selectedServer = null;

        for (Server server : servers) {
            if (server.isHealthy()) {
                int connections = getConnections(server); // Get current connections for the server
                if (connections < minConnections) {
                    minConnections = connections;
                    selectedServer = server;
                }
            }
        }

        if (selectedServer == null) {
            throw new IllegalStateException("No healthy servers available");
        }
        return selectedServer;
    }

    private int getConnections(Server server) {
        // Perform logic to get current connections for the server
        return 0; // Placeholder for connections count
    }
}

// Main Class
public class LoadBalancerApp {
    public static void main(String[] args) {
        // Create servers
        Server server1 = new Server("server1");
        Server server2 = new Server("server2");

        // Create load balancer
        LoadBalancer loadBalancer = LoadBalancer.getInstance();
        loadBalancer.addServer(server1);
        loadBalancer.addServer(server2);

        // Set load balancing strategy
        LoadBalancingStrategy roundRobinStrategy = new RoundRobinStrategy();
        loadBalancer.setLoadBalancingStrategy(roundRobinStrategy);

        // Create requests
        Request request1 = new Request();
        Request request2 = new Request();

        // Get server for request1
        Server selectedServer1 = loadBalancer.getServer(request1);
        System.out.println("Selected server for request1: " + selectedServer1.getServerId());

        // Get server for request2
        Server selectedServer2 = loadBalancer.getServer(request2);
        System.out.println("Selected server for request2: " + selectedServer2.getServerId());
    }
}
